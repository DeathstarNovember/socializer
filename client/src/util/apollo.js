import ApolloClient from "apollo-client";
import { InMemoryCache } from "apollo-cache-inmemory";
import * as AbsintheSocket from "@absinthe/socket";
import { createAbsintheSocketLink } from "@absinthe/socket-apollo-link";
import { Socket as PhoenixSocket, LongPoll } from "phoenix";
import { createHttpLink } from "apollo-link-http";
import { hasSubscription } from "@jumpn/utils-graphql";
import { split } from "apollo-link";
import { setContext } from "apollo-link-context";
import Cookies from "js-cookie";

const HTTP_URI =
  process.env.NODE_ENV === "production"
    ? "https://brisk-hospitable-indianelephant.gigalixirapp.com"
    : "http://localhost:4000";

const WS_URI =
  process.env.NODE_ENV === "production"
    ? "wss://brisk-hospitable-indianelephant.gigalixirapp.com/socket"
    : "ws://localhost:4000/socket";

function createPhoenixSocket() {
  // Socket with full fallback to LongPoll
  // via https://elixirforum.com/t/fall-back-to-longpoll-when-websocket-fails/23894
  const socket = new PhoenixSocket(WS_URI, {
    params: () => {
      if (Cookies.get("token")) {
        return { token: Cookies.get("token") };
      } else {
        return {};
      }
    },
  });
  socket.onError(() => {
    if (navigator.onLine) {
      const isWebSocket = socket.transport === window.WebSocket;
      console.error(
        `Error connecting using ${isWebSocket ? "WebSocket" : "LongPoll"}`,
      );
      if (isWebSocket) socket.transport = LongPoll;
      else if (window.WebSocket) socket.transport = window.WebSocket;
    }
  });
  return socket;
}

export const createClient = ({ ssr, req, fetch, tokenCookie } = {}) => {
  let link = createHttpLink({ uri: HTTP_URI, fetch });
  let absintheSocket;

  if (!ssr) {
    absintheSocket = AbsintheSocket.create(createPhoenixSocket());
    const socketLink = createAbsintheSocketLink(absintheSocket);

    link = split(
      (operation) => hasSubscription(operation.query),
      socketLink,
      link,
    );
  }

  const authLink = setContext((_, { headers }) => {
    // Get the authentication token from the cookie if it exists.
    const token = tokenCookie || Cookies.get("token");

    // Return the headers to the context so httpLink can read them.
    return {
      headers: {
        ...headers,
        authorization: token ? `Bearer ${token}` : "",
      },
    };
  });

  const cache = new InMemoryCache();
  if (!ssr) {
    cache.restore(window.__APOLLO_STATE__);
  }

  const client = new ApolloClient({
    connectToDevTools: !ssr,
    ssrMode: ssr,
    link: authLink.concat(link),
    cache,
  });
  return { client, absintheSocket };
};

export const refreshSocket = (socket) => {
  // Close the connection to force a reconnection with the
  // new token parameter.
  socket.phoenixSocket.conn && socket.phoenixSocket.conn.close();
};
