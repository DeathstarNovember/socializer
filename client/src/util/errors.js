import { camelizeKeys } from "humps";

export const errorHash = ({ graphQLErrors } = {}) => {
  if (!Array.isArray(graphQLErrors)) return {};
  return camelizeKeys(
    graphQLErrors.reduce(
      (hash, error) => Object.assign(hash, { [error.field]: error.message }),
      {},
    ),
  );
};
