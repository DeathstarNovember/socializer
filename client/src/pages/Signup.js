import React, { useContext, useState } from "react";
import { useMutation } from "@apollo/react-hooks";
import { Button, Col, Container, Form, Row } from "react-bootstrap";
import { Helmet } from "react-helmet";
import gql from "graphql-tag";
import { Redirect } from "react-router-dom";
import renderIf from "render-if";
import { AuthContext } from "util/context";
import { errorHash } from "util/errors";

export const SIGNUP = gql`
  mutation Signup($name: String!, $email: String!, $password: String!) {
    signUp(name: $name, email: $email, password: $password) {
      id
      token
    }
  }
`;

const useErrors = () => {
  const [errors, setErrors] = useState({});

  const clearError = (field) =>
    setErrors(Object.assign(errors, { [field]: null }));

  return [errors, setErrors, clearError];
};

const Signup = () => {
  const { token, setAuth } = useContext(AuthContext);
  const [errors, setErrors, clearError] = useErrors();
  const [name, setName] = useState("");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [signUp, { data, loading }] = useMutation(SIGNUP, {
    onError: (error) => setErrors(errorHash(error)),
  });

  if (data) {
    const {
      signUp: { id, token },
    } = data;
    setAuth({ id, token });
  }

  if (token) {
    return <Redirect to="/" />;
  }

  return (
    <>
      <Helmet>
        <title>Socializer | Sign up</title>
        <meta property="og:title" content="Socializer | Sign up" />
      </Helmet>

      <Container>
        <Row>
          <Col md={6} xs={12}>
            <Form
              onSubmit={(e) => {
                e.preventDefault();
                signUp({
                  variables: {
                    name,
                    email,
                    password,
                  },
                });
              }}
            >
              <Form.Group controlId="formName">
                <Form.Label>Name</Form.Label>
                <Form.Control
                  placeholder="Joe Smith"
                  value={name}
                  onChange={(e) => {
                    setName(e.target.value);
                    clearError("name");
                  }}
                  isInvalid={Boolean(errors.name)}
                />
                {renderIf(errors.name)(
                  <Form.Control.Feedback type="invalid">
                    {errors.name}
                  </Form.Control.Feedback>,
                )}
              </Form.Group>

              <Form.Group controlId="formEmail">
                <Form.Label>Email address</Form.Label>
                <Form.Control
                  type="email"
                  placeholder="you@gmail.com"
                  value={email}
                  onChange={(e) => {
                    setEmail(e.target.value);
                    clearError("email");
                  }}
                  isInvalid={Boolean(errors.email)}
                />
                {renderIf(errors.email)(
                  <Form.Control.Feedback type="invalid">
                    {errors.email}
                  </Form.Control.Feedback>,
                )}
              </Form.Group>

              <Form.Group controlId="formPassword">
                <Form.Label>Password</Form.Label>
                <Form.Control
                  type="password"
                  placeholder="Password"
                  value={password}
                  onChange={(e) => {
                    setPassword(e.target.value);
                    clearError("password");
                  }}
                  isInvalid={Boolean(errors.password)}
                />
                {renderIf(errors.password)(
                  <Form.Control.Feedback type="invalid">
                    {errors.password}
                  </Form.Control.Feedback>,
                )}
              </Form.Group>

              <Button variant="primary" type="submit" disabled={loading}>
                {loading ? "Signing up..." : "Sign up"}
              </Button>
            </Form>
          </Col>
        </Row>
      </Container>
    </>
  );
};

export default Signup;
