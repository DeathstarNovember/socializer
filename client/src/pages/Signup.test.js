import React from "react";
import { render, fireEvent, wait } from "@testing-library/react";
import { MockedProvider } from "@apollo/react-testing";
import { MemoryRouter } from "react-router-dom";
import { AuthContext } from "util/context";
import Signup, { SIGNUP } from "./Signup";

jest.mock("react-router-dom", () => ({
  MemoryRouter: jest.requireActual("react-router-dom").MemoryRouter,
  Redirect: () => <div className="redirect" />,
}));

describe("Signup", () => {
  it("renders correctly when authenticated", () => {
    const { container } = render(
      <MemoryRouter>
        <AuthContext.Provider value={{ token: "abc", setAuth: jest.fn() }}>
          <MockedProvider mocks={[]} addTypename={false}>
            <Signup />
          </MockedProvider>
        </AuthContext.Provider>
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly when unauthenticated", () => {
    const { container } = render(
      <MemoryRouter>
        <AuthContext.Provider value={{ setAuth: jest.fn() }}>
          <MockedProvider mocks={[]} addTypename={false}>
            <Signup />
          </MockedProvider>
        </AuthContext.Provider>
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  it("signs up", async () => {
    const mocks = [
      {
        request: {
          query: SIGNUP,
          variables: {
            name: "John Lvh",
            email: "john@lvh.me",
            password: "password",
          },
        },
        result: { data: { signUp: { id: 123, token: "abc" } } },
      },
    ];
    const setAuth = jest.fn();
    const { getByLabelText, getByText } = render(
      <MemoryRouter>
        <AuthContext.Provider value={{ setAuth }}>
          <MockedProvider mocks={mocks} addTypename={false}>
            <Signup />
          </MockedProvider>
        </AuthContext.Provider>
      </MemoryRouter>,
    );
    fireEvent.change(getByLabelText("Name"), {
      target: { value: "John Lvh" },
    });
    fireEvent.change(getByLabelText("Email address"), {
      target: { value: "john@lvh.me" },
    });
    fireEvent.change(getByLabelText("Password"), {
      target: { value: "password" },
    });
    fireEvent.click(getByText("Sign up"));
    await wait(() => expect(setAuth).toBeCalledWith({ id: 123, token: "abc" }));
  });
});
