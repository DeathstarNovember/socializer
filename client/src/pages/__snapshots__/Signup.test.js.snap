// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup renders correctly when authenticated 1`] = `
<div>
  <div
    class="redirect"
  />
</div>
`;

exports[`Signup renders correctly when unauthenticated 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 col-12"
      >
        <form
          class=""
        >
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="formName"
            >
              Name
            </label>
            <input
              class="form-control"
              id="formName"
              placeholder="Joe Smith"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="formEmail"
            >
              Email address
            </label>
            <input
              class="form-control"
              id="formEmail"
              placeholder="you@gmail.com"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="formPassword"
            >
              Password
            </label>
            <input
              class="form-control"
              id="formPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <button
            class="btn btn-primary"
            type="submit"
          >
            Sign up
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
